rust/hyperlane-base/src/metrics/agent_metrics.rs

```diff
@@ -0,0 +1,119 @@
+use std::time::Duration;
+
+use derive_builder::Builder;
+use derive_new::new;
+use eyre::Result;
+use hyperlane_core::metrics::agent::u256_as_scaled_f64;
+use hyperlane_core::{metrics::agent::AgentMetricsFetcher, HyperlaneDomain};
+use maplit::hashmap;
+use prometheus::GaugeVec;
+use tokio::time::MissedTickBehavior;
+use tracing::{trace, warn};
+
+use crate::CoreMetrics;
+
+/// Expected label names for the `wallet_balance` metric.
+pub const WALLET_BALANCE_LABELS: &[&str] = &[
+    "chain",
+    "wallet_address",
+    "wallet_name",
+    "token_address",
+    "token_symbol",
+    "token_name",
+];
+/// Help string for the metric.
+pub const WALLET_BALANCE_HELP: &str =
+    "Current native token balance for the wallet addresses in the `wallets` set";
+
+/// Agent-specific metrics
+#[derive(Clone, Builder)]
+pub struct Metrics {
+    /// Current balance of native tokens for the
+    /// wallet address.
+    /// - `chain`: the chain name (or chain ID if the name is unknown) of the
+    ///   chain the tx occurred on.
+    /// - `wallet_address`: Address of the wallet holding the funds.
+    /// - `wallet_name`: Name of the address holding the funds.
+    /// - `token_address`: Address of the token.
+    /// - `token_symbol`: Symbol of the token.
+    /// - `token_name`: Full name of the token.
+    #[builder(setter(into, strip_option), default)]
+    wallet_balance: Option<GaugeVec>,
+}
+
+pub(crate) fn create_agent_metrics(metrics: &CoreMetrics) -> Result<Metrics> {
+    Ok(MetricsBuilder::default()
+        .wallet_balance(metrics.new_gauge(
+            "wallet_balance",
+            WALLET_BALANCE_HELP,
+            WALLET_BALANCE_LABELS,
+        )?)
+        .build()?)
+}
+
+/// Configuration for the prometheus middleware. This can be loaded via serde.
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(tag = "type", rename_all = "camelCase"))]
+pub struct AgentMetricsConf {
+    /// The account to track
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub address: Option<String>,
+
+    /// Information about the chain this metric is for
+    pub domain: HyperlaneDomain,
+
+    /// Name of the agent the metrics are about
+    pub name: String,
+}
+
+/// Utility struct to update agent metrics
+#[derive(new)]
+pub struct AgentMetrics {
```

Comment:
The name implies to me imo that there's just a single one of these for each agent, but really it seems like this is tied to a chain that the agent is using. Wondering if a name like `AgentChainMetrics` or something would be more appropriate? open to other ideas

actually looking a bit deeper I think the naming of `Metrics` (above) / `AgentMetrics` and how they're used is confusing

e.g. `create_agent_metrics` above creates `Metrics`, but in the `from_settings` function when creating an agent, `Metrics` is often imported as `AgentMetrics` and called `agent_metrics`, but this is different from `AgentMetrics` here. i think consistent naming would help